### ISSUE TITLE
(sidenav) elements with `position: fixed` inside `md-sidenav-layout`

### ISSUE THREAD

**Issue by fxck** [angular/material2#998]
Because of `transform: translate3d(0, 0, 0)` on `md-sidenav-layout` any element with `position: fixed` inside it will act as if it had `position: absolute` or `static` instead. It's an [expected behaviour](http://stackoverflow.com/a/15256339/301596), but it's annoying nevertheless.

Could we either remove that `transform` from `md-sidenav-layout`(it's only there because of performance I suppose), or add more examples to `Portal` and mention it as a workaround in docs, since portal could probably solve this by basically putting the element that needs to be fixed next to the `md-sidenav-layout` in the dom tree.

cc @hansl @jelbourn

**Comment by fxck** [comment]
hello?

**Comment by fxck** [comment]
```typescript

import {
  Component,
  Directive,
  TemplateRef,
  ViewContainerRef,
  ViewChild,
  Input,
  OnDestroy,
  AfterViewInit
} from '@angular/core';

import {
  TemplatePortal,
  OverlayState,
  OverlayRef,
  Overlay
} from '@angular2-material/core';

@Directive({ selector: '[displacerPortal]' })
export class DisplacerPortalDirective extends TemplatePortal {
  constructor(templateRef: TemplateRef<any>, viewContainerRef: ViewContainerRef) {
    super(templateRef, viewContainerRef);
  }
}

@Component({
  selector: 'fnls-displacer',
  styles: [ require('./displacer.component.scss') ],
  template: require('./displacer.component.html')
})
export class DisplacerComponent implements OnDestroy {

  private _config = new OverlayState();

  @ViewChild(DisplacerPortalDirective)
  private _portal: DisplacerPortalDirective;

  private _overlayRef: OverlayRef = undefined;

  constructor(private _overlay: Overlay) {}

  public ngOnDestroy() {
    this._overlayRef.detach();
  }

  public ngAfterViewInit() {
    console.log(this._portal);
    this._overlayRef = this._overlay.create(this._config);
    this._overlayRef.attach(this._portal);
  }
}

```

**Comment by fxck** [comment]
@jelbourn this is still a thing, can you label it please?

**Comment by jelbourn** [comment]
We need to discuss whether we consider this a bug or not.

**Comment by cortopy** [comment]
In my opinion this is a bug insofar as it is the equivalent of putting a ban on fixed elements. I am now trying to create a fab button on the bottom right corner of the screen. Because of this issue I'm unable to do so.

**Comment by bergben** [comment]
Is the workaround from @fxck still working? 
I am getting `EXCEPTION: Must provide a portal to attach`...

**Comment by mmalerba** [comment]
I recommend just absolutely positioning the fab outside the scrollable region: http://plnkr.co/edit/VItn3Y2AaTzpzIykTTyH?p=preview

**Comment by cortopy** [comment]
that's clever @mmalerba but it wouldn't work in my case. My button is inside a router outlet as the fab is not always needed. Your solution would work great for cases where fixed elements should be visible in every view though.

Also, it would mean a lot of boilerplate to pass events to other components within the scrollable area

**Comment by cortopy** [comment]
@bergben it is. I'm using material alpha 10 and it still works. It took me a bit to get it working though. Could you post what you have?

**Comment by bergben** [comment]
@cortopy 
It was kinda late yesterday, the problem was that I simply forgot to import the Directive as well 😄 
Got some other problem now though that the sidebar is not opening / visible / whatever... 
What I got now is this: 
```html```
<md-sidenav-layout [ngClass]="{'administration' : isLoggedIn}">
    <md-sidenav #sideNav mode="push" [hidden]="!isLoggedIn" opened="true">
        <noe-admin-sidenav>
        </noe-admin-sidenav>
    </md-sidenav>
    <fnls-displacer [sideNav]="sideNav">
            <noe-admin-topnav [sideNav]="sideNav" *ngIf="isLoggedIn">
            </noe-admin-topnav>
            <div class="content-wrapper">
                <router-outlet></router-outlet>
                <noe-footer></noe-footer>
            </div
    </fnls-displacer>
</md-sidenav-layout>
```
How exactly do I have to wrap the content with the displacer to get it to work? 
I also had to add this css to make the fixed content work:
```css```
.md-overlay-pane{
    width:100%;
}
.md-overlay-container{
    position:absolute !important;
}
```

**Comment by bergben** [comment]
Still couldn't resolve this. Seems like this wasn't a problem with the angular 1 version of the sidenav... 
Problem is, I don't only have one button that is fixed but the whole content is fixed for a certain time as I am using https://github.com/bergben/ng2-scrollimate to apply animations on the fly during scroll. So if I displace that all then there is nothing left anymore in md-sidenav-content and the sidenav itself doesn't work anymore...

**Comment by mmalerba** [comment]
#1966 adds a sidenav with fab example

**Comment by fxck** [comment]
Having a scroll inside content rather than on body brings a lot problems with UX and with fixed elements(various cases when fixed element is actually overlaying the scrollbar, or regions that appear part of scroll but are not, so you are scrolling on them and the page doesn't scroll), als there are potential problems / extra set up on phones for smooth scrolling required, potential double scrollbars... meh. All this because of a unnecessary `transform: translate3d(0, 0, 0)`. 👍

**Comment by mmalerba** [comment]
I do think separate scrolling areas should be the recommended solution for this. If I have a lot of content in my sidenav I probably want it to scroll independently from my content area. However I do think it currently requires a bit too much work to set up, I would like to make it easier if possible. I'm also not opposed to removing the translate3d if it really is superfluous (I need to investigate) so that people can use the fixed positioning solution if they prefer.

@jelbourn Do you have any opinions on this?

**Comment by bergben** [comment]
@mmalerba This solution only works if you have relatively simple fixed content, if you have it spanning over the whole page or stuff like that it just gets too complicated. What I find interesting though is that this problem didn't seem to come up in the Angular 1 version of the sidenav - as I was searching through the issues there. 
The translate3d could be replaced with a simple animation and basic distance property to fix this problem for good... 
I don't know though why the translate3d breaks a css position in the first place, this is surely a misbehavior of CSS in the first place IMO, but that's another topic.

**Comment by johnl242** [comment]
I am using angular 2 material design, pls give a solution base on angular 2 material design and not material 1

**Comment by cortopy** [comment]
@bergben I'm not sure I understand your template.

Simplifying a lot this is what I have in `app.component.html`:

```html
<md-sidenav-layout>
    <md-sidenav #start mode="side">
      <my-sidenav></my-sidenav>
    </md-sidenav>

    <router-outlet></router-outlet>
 </md-sidenav-layout>
```

And then in a template for a component that gets loaded by the router I have:

```html
<fnls-displacer>
      <button></button>
</fnls-displacer>
```

**Comment by johnl242** [comment]
fab button not being fixed on button right position while scrolling

**Comment by fxck** [comment]
Yeah you got that right @cortopy there are no inputs on displacer itself, it's literally just "take whatever is inside me and append it to body". It doesn't break bindings or anything, so it's as if you were wrapping it in a div. What I think he's doing is wrapping the whole router-outlet inside it, which I'm not surprised at all that doesn't work. It's really intended for buttons/modals/whatever you need appended to body from *inside* the router outlet.

cc @bergben

**Comment by johnl242** [comment]
I have the fab button inside a component and this component is being called
from router-outlet and the router-outlet is inside the side-nav-layout as u
have described. After doing every thing the fab scrolls while scrolling
data in the component which is not what I need.

On Nov 29, 2016 10:01 PM, "Aleš" <notifications@github.com> wrote:

Yeah you got that right @cortopy <https://github.com/cortopy> there are no
inputs on displacer itself, it's literally just "take whatever is inside me
and append it to body". It doesn't break bindings or anything, so it's as
if you were wrapping it in a div. What I think he's doing is wrapping the
whole router-outlet inside it, which I'm not surprised at all that doesn't
work. It's really intended for buttons/modals/whatever you need appended to
body from *inside* the router outlet.

cc @bergben <https://github.com/bergben>

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/angular/material2/issues/998#issuecomment-263621269>,
or mute the thread

**Comment by cortopy** [comment]
@johnl242 don't forget to style your button!! The displacer doesn't do anything in terms of styling or positioning. This is what I have

```sass
  bottom: 1rem
  position: fixed
  right: 1rem
  transform: translateZ(0)
```

**Comment by johnl242** [comment]
It's positing at fixed bottom right  but scrolling while data being scrolled

**Comment by bergben** [comment]
@johnl242 I think it would be better if you post some code to see what you have so that people can actually help you. 

@cortopy @fxck  Thanks but the displacer is no solution for me as the content is only fixed during a certain period of time and the rest of the time it is normally positioned inside some other content. As said I am using ng2-scrollimate to create some scrolling effects which also include having an element fixed for a certain period of scroll. Porting that element to the body does not offer a solution as it breaks the whole layout. So I guess I'll have to wait for a solution that does not use `transform: translate3d`

**Comment by bradysterling** [comment]
pretty please remove `transform: translate3d` from the `.md-sidenav-content` class?

**Comment by Carlos-Montiel** [comment]
So, what's the solution?

**Comment by Splaktar** [comment]
`.mat-tab-body-content` of `md-tab-body` has `transform: translate3d(0px, 0px, 0px);` as well. Thus trying to add a FAB inside of tab container is also a problem.

**Comment by joonaspi** [comment]
Any update on this? I would need to have a FAB button floating while user scrolls down the page.

**Comment by alexichepura** [comment]
CSS transform on .mat-sidenav-content creates a new local coordinate system.
For now I use this workaround.
```css
.mat-sidenav-content {
    transform: none !important;
}
```

**Comment by Defmetalhead** [comment]
Currently having to override globally

```
.mat-sidenav-container, .mat-sidenav-content {
    transform: none !important;
    transition: none !important;
}

```

**Comment by radoslavpetranov** [comment]
@Defmetalhead removing the transition screws up the sidebar content animation when you toggle the menu. In my case using just

```
.mat-sidenav-container, .mat-sidenav-content {
    transform: none !important;
}
```

was all I needed to get the `position: fixed` to work. Just a heads up to anyone who just copy/pastes this without noticing the side effect.

**Comment by Defmetalhead** [comment]
@radoslavpetranov I do not experience the issue with the sidebar animation, but you are correct in the fact that only "transform" needs to be overwritten.

**Comment by gatimus** [comment]
Just a note this is also causing change in behavior in other libraries as well. [http://wijmo.com/topic/how-to-prevent-window-scroll-to-top-on-wijmo-flexgrid-click-event/](http://wijmo.com/topic/how-to-prevent-window-scroll-to-top-on-wijmo-flexgrid-click-event/)

**Comment by ArthurianX** [comment]
It is actually breaking bindings, a form submit `type=submit` button.

If I manually submit through a `(click)` it works, but the submit itself, it gets it out of context.

**Comment by ncrawlins** [comment]
I use a third party charting library in the content and the transform breaks all their mouseover tooltips in certain container structures..  It also causes full page flickers in IE 11 when their tooltips are active.  Turning off the 3d transform fixes all these things.  What does it do? Is it necessary?

**Comment by mackelito** [comment]
using transform will (on some devices) trigger to use hardware acceleration.. feels like a good incentive to keep it.. ?

**Comment by mugwag** [comment]
+1

**Comment by jvanhall** [comment]
With the latest release the component @fxck built no longer works due to the TemplatePortal changes.  Can we get an updated version as patch until this is resolved.  Also, just removing transforms via CSS no longer seems to be a valid work around.

**Comment by ncrawlins** [comment]
the transform fix still works for me but they changed the class names from .mat-sidenav-* to .mat-drawer-* so you probably have to update your css

**Comment by jvanhall** [comment]
Yeah looks like I missed one of the .mat-drawer classes.  Working now.

**Comment by WaterBleu** [comment]
Im surprised we still have to do workaround at this point.
this is like a year old bug

**Comment by CDDelta** [comment]
@WaterBleu see #6712

**Comment by WaterBleu** [comment]
@CDDelta nice! now just have to wait for the approval

**Comment by MatthewS2077** [comment]
@CDDelta Excellent!

**Comment by mackelito** [comment]
aaaah maaaaan.. just I did this like 3 weeks ago! 😆 
But still.. Great work 👊 😄

**Comment by karser** [comment]
Great that it'll be fixed soon.
Meanwhile after updating to 2.0.0-beta.10:
```
.mat-sidenav-content,
.mat-sidenav-container,
.mat-drawer-content {
    transform: none !important;
}
```

**Comment by sijithc40** [comment]
I given this
.mat-sidenav-content,
.mat-sidenav-container,
.mat-drawer-content {
    transform: none !important;
}

But nab push is not working now

**Comment by gatimus** [comment]
@sijithc40 It won't because "push" mode uses `transform`.

**Comment by Sangeethaself** [comment]
Material design light with version v1.0.1

After long searching and trying , problem solved by following code

 .mat-sidenav-container,
   md-sidenav-container div[style],
  .mat-sidenav-content{
    transform: none !important;
    -webkit-transform: none !important;
  }

**Comment by rahulasharm** [comment]
I am using angular 2, so what should I do for this issue.....

**Comment by edgarMejia** [comment]
Simply define width margin at mat-sidenav-content in styles.css file:
```
mat-sidenav-content {
    width: 100vw;
}
```
